loopy/transform/iname.py

```diff
@@ -157,7 +157,62 @@ def map_variable(self, expr, expn_state):
             return super().map_variable(expr, expn_state)
 
 
-def _split_iname_backend(kernel, split_iname,
+def _split_iname_in_set(s, iname_to_split, inner_iname, outer_iname, fixed_length,
+        fixed_length_is_inner, split_iname_should_remain):
+    var_dict = s.get_var_dict()
+
+    if iname_to_split not in var_dict:
+        return s
+
+    orig_dim_type, _ = var_dict[iname_to_split]
+    assert orig_dim_type == dim_type.set
+    del orig_dim_type
+
+    # NB: dup_iname_to_split is not a globally valid identifier: only uniqure
+    # wrt the set s.
+    from pytools import generate_unique_names
+    for dup_iname_to_split in generate_unique_names(f"dup_{iname_to_split}"):
+        if dup_iname_to_split not in var_dict:
+            break
+
+    from loopy.isl_helpers import duplicate_axes
+    s = duplicate_axes(s, (iname_to_split,), (dup_iname_to_split,))
+
+    outer_var_nr = s.dim(dim_type.set)
+    inner_var_nr = s.dim(dim_type.set)+1
+
+    s = s.add_dims(dim_type.set, 2)
+    s = s.set_dim_name(dim_type.set, outer_var_nr, outer_iname)
+    s = s.set_dim_name(dim_type.set, inner_var_nr, inner_iname)
+
+    from loopy.isl_helpers import make_slab
+
+    if fixed_length_is_inner:
+        fixed_iname, var_length_iname = inner_iname, outer_iname
+    else:
+        fixed_iname, var_length_iname = outer_iname, inner_iname
+
+    space = s.get_space()
+    s = s & (
+            make_slab(space, fixed_iname, 0, fixed_length)
+            # name = fixed_iname + fixed_length*var_length_iname
+            .add_constraint(isl.Constraint.eq_from_names(
+                space, {
+                    dup_iname_to_split: 1,
+                    fixed_iname: -1,
+                    var_length_iname: -fixed_length})))
+
+    _, dup_name_idx = space.get_var_dict()[dup_iname_to_split]
+    s = s.project_out(dim_type.set, dup_name_idx, 1)
+
+    if split_iname_should_remain:
```

Comment:
How about always letting the iname_to_split be a part of the domain and calling a `remove_unused_inames` at the end. Slightly simplifies the logic, IMO.

Reply:
Good idea. 8e4d3ebe50cfa91484cfcfd0cae8a8701c488811